### ISSUE TITLE
add hex_to_hash() function

### ISSUE THREAD

**Issue by Dragosha** [defold/defold#7664]
related to this post:

> We should add a function to go from a hash to something serialisable and back again. We already have hash_to_hex() so we need a hex_to_hash().

https://forum.defold.com/t/how-to-save-a-huge-table-with-subtables/68749/13?u=dragosha

In short, this function is needed to safely store a hashed value in json and convert it back to hash when the json is loaded.

**Comment by JCash** [comment]
Can I ask _what_ is being saved here?

My main concern with storing a hash is that it is an engine "internal" data format.
If we were to change that format in the future, then the saved file might not be possible to use in a meaningful way, as there will be no way to convert from old hash to new hash.

I think it would be much better to store the original strings, depending on how they're created.
But, as I said I'd need to know more about the keys first.

In short, I'd consider it a power user feature, and I'd recommend also storing your own version number in the file (to detect if we change hash algorithm)

**Comment by Dragosha** [comment]
> If we were to change that format in the future, then the saved file might not be possible to use in a meaningful way, as there will be no way to convert from old hash to new hash.
I absolutely agree with this point. At the same time, the request for this function started because hash is the only internal format for storing writeable values in the go.property script (we can't store strings). 
My recent example is this: a level has door objects that lead to other levels. The hash in this case is a key in a table that stores the information already needed to load the level (collection factory address, name, etc.).
To save current level name in json save I just had to add another string key and make a hash -> id_key matching table.

Same thing happens when you need to encode some other meaningful key in an object, e.g. in the item object specify what kind of object it is, apple or super-duper power sword. Again, this is a hash key in the item table. Sooner or later the character's inventory has to be saved in json and again the question arises how. I used to add another id. But in that case when loading a save you have to do a full pass through the item table looking for the right id to find out the hash key and fill in the character's inventory again.

![2023-05-29_13-30-38](https://github.com/defold/defold/assets/842807/7e8c74d6-8268-4cfb-82a5-800589db85c3)

Of course you can store such keys in a table as a number, but this is not very convenient when creating levels in the editor.
If the script had the ability to store strings (with a strong mention in the manual not to just use them for everything in a row) the question would not arise. 
Surely there are other cases as well.

**Comment by JCash** [comment]
> If the script had the ability to store strings

Not sure what you mean by this?
We can use strings as keys.

**Comment by Dragosha** [comment]
> > If the script had the ability to store strings
> 
> Not sure what you mean by this? We can use strings as keys.

I mean script  `go.property()` method. 

Actually this whole story is about creating game levels inside the Defold editor. We need a binding between the object and a table with parameters, now we can have either number or hash as a key. If we use a human-readable hash as a key, then we cannot bind it directly to the storage (e.g. in the cloud, where json is used).

**Comment by rigo128** [comment]
I encountered the same problem with the same use cases.
